### PR TITLE
Title : Add new template path for Jinja service 

### PR DESCRIPTION
**Pull request by mithunbharadwaj**

### CODE REVIEW

File: asabiris/orchestration/render.py [hunk 1]
```diff
@@ -17,14 +17,15 @@ def __init__(self, app):
 		self.JinjaService = app.get_service("JinjaService")
 		self.HtmlToPdfService = app.get_service("HtmlToPdfService")
 		self.MarkdownToHTMLService = app.get_service("MarkdownToHTMLService")
-
+		self.TempPath = "/Template/render"
```
Comment (ateska) [inline asabiris/orchestration/render.py:20]: - [ ] `/Templates` should be configurable (`asab.config`) and used everywhere.
- [ ] `/Templates/Generic`
- [ ] `/Templates` (not `/Template`)

File: asabiris/orchestration/render.py [hunk 2]
```diff
@@ -17,14 +17,15 @@ def __init__(self, app):
...
 
 	async def render(self, template, params):
 		"""
 		This method renders templates based on the depending on the
 		extension of template. Returns the html/pdf.
 		"""
-		html = await self.JinjaService.format(template, params)
+		template_path = self.TempPath + template
```
Comment (ateska) [inline asabiris/orchestration/render.py:27]: Check that `template` starts with `/`, raise the exceptions if not.

File: asabiris/orchestration/sendmail.py [hunk 3]
```diff
@@ -23,7 +23,7 @@ def __init__(self, app):
 
 		# output
 		self.SmtpService = app.get_service("SmtpService")
-
+		self.TempPath = "/Templates/email"
```
Comment (ateska) [inline asabiris/orchestration/sendmail.py:26]: - [ ] dtto ^
- [ ] `/Templates/Email`

File: asabiris/orchestration/sendmail.py [hunk 4]
```diff
@@ -71,7 +71,8 @@ async def send_mail(
 
 					# get file-name of the attachment
 					file_name = self.get_file_name(a)
-					jinja_output, result = await self.render(template, params)
+					template_path = self.TempPath + template
```
Comment (ateska) [inline asabiris/orchestration/sendmail.py:74]: dtto

File: asabiris/orchestration/sendslack.py [hunk 5]
```diff
@@ -19,7 +19,8 @@ def __init__(self, app):
 		self.JinjaService = app.get_service("JinjaService")
 		# output
 		self.SlackOutputService = app.get_service("SlackOutputService")
-
+		# location of slack templates
+		self.TempPath = "/Templates/slack"
```
Comment (ateska) [inline asabiris/orchestration/sendslack.py:23]: - [ ] Dtto
- [ ] `/Templates/Slack`

File: asabiris/orchestration/sendslack.py [hunk 6]
```diff
@@ -31,7 +32,7 @@ async def send_to_slack(self, msg):
 		body = msg['body']
 		# if params no provided pass empty params
 		body["params"] = body.get("params", {})
-
+		body['template'] = self.TempPath + body['template']
```
Comment (ateska) [inline asabiris/orchestration/sendslack.py:35]: Dtto

**Review by ateska — CHANGES_REQUESTED** [review]
Please check with the team if the top level entry in the library (`/Templates`) should be plural or singular.

File: asabiris/orchestration/render.py (hunk 1 above)
Comment (mithunbharadwaj) [reply asabiris/orchestration/render.py:20]: @ateska Sure , I will

**Review by ateska — APPROVED** [review]

**Comment by eliska-n** [comment]
```
23-Jan-2023 11:01:22.752860 ERROR asab.web [sd uuid="1c308df6-b04c-4774-86b5-062c3e3917f1" path="/send_mail" status="500" Host="localhost:8082" Accept="*/*" Accept-Encoding="gzip, deflate" User-Agent="Python/3.8 aiohttp/3.8.1" Content-Length="11767" Content-Type="application/json"] Exception when handling web request
Traceback (most recent call last):
  File "/home/eliska/teskalabs/asab/asab/web/rest/json.py", line 82, in JsonExceptionMiddleware
    response = await handler(request)
  File "/home/eliska/teskalabs/asab/asab/web/rest/json.py", line 288, in validator
    return await func(*args, **kwargs)
  File "/home/eliska/teskalabs/asab-iris/asabiris/handlers/webhandler.py", line 95, in send_mail
    await self.App.SendMailOrchestrator.send_mail(
  File "/home/eliska/teskalabs/asab-iris/asabiris/orchestration/sendmail.py", line 69, in send_mail
    if not template.startswith("/Template"):
AttributeError: 'NoneType' object has no attribute 'startswith'
```

This doesn't look like a sweet warning :)

**Comment by eliska-n** [comment]
Mithun, please, add the base64 attachment case to your testing - unittests, postman collection - whatever you use to evaluate your code. It is vital for bsquery. It doesn't need to be an export file. You can generate base64 picture online, for example.

File: asabiris/orchestration/sendslack.py [hunk 7]
```diff
@@ -30,8 +30,9 @@ async def send_to_slack(self, msg):
 
 		body = msg['body']
 		# if params no provided pass empty params
+		if not body['template'].startswith("/Template"):
```
[the remark this answers is not in the document]
Comment (eliska-n) [reply asabiris/orchestration/sendslack.py:33]: it is "/Templates"

[the remark this answers is not in the document]
Comment (eliska-n) [reply asabiris/orchestration/sendslack.py:33]: @ateska - I would consider making this configurable as this microservice is opensource. Having library items hardcoded here might seem confusing for anyone outside of TeskaLabs. But I would keep this for some other MR.

File: asabiris/orchestration/sendmail.py [hunk 8]
```diff
@@ -53,7 +52,7 @@ async def send_mail(
 		# Render a body
 		body_html, email_subject_body = await self.render(body_template, body_params)
```
Comment (eliska-n) [inline asabiris/orchestration/sendmail.py:53]: Mithun, you changed the way how you look into the library only when handling attachments, but you must do the same for body template as well.

File: asabiris/orchestration/render.py [hunk 9]
```diff
@@ -18,12 +18,13 @@ def __init__(self, app):
 		self.HtmlToPdfService = app.get_service("HtmlToPdfService")
 		self.MarkdownToHTMLService = app.get_service("MarkdownToHTMLService")
 
-
 	async def render(self, template, params):
 		"""
 		This method renders templates based on the depending on the
 		extension of template. Returns the html/pdf.
 		"""
+		if not template.startswith("/Template"):
```
[the remark this answers is not in the document]
Comment (eliska-n) [reply asabiris/orchestration/render.py:26]: And it is not the whole story. We agreed on this functionality:
```
- how to read template name:
	- primarily use absolute path - starts with "/"
	- if absolute path is used, check it start with "/Templates"
	- if it is not absolute path, it is file name - search for that file in Templates directory
```

**Review by eliska-n — CHANGES_REQUESTED** [review]

File: asabiris/orchestration/sendmail.py [hunk 10]
```diff
@@ -53,7 +52,7 @@ async def send_mail(
...
 
-		if email_subject is None:
```
Comment (eliska-n) [inline asabiris/orchestration/sendmail.py:56]: This line was right.
If Subject is not in the request, use the one from template.
If Subject is not in the template, use the default from the configuration.

Comment (mithunbharadwaj) [reply asabiris/orchestration/sendmail.py:56]: @eliska-n What happens if subject is in the request and template?

File: asabiris/orchestration/render.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (mithunbharadwaj) [reply asabiris/orchestration/render.py]: @eliska-n what I know from WS and from DEMO

Templates can be generic or specifics.

All the generic Templates should sit inside /Templates/Generic/

Specific templates:

e-mail : /Templates/Mails/

Slack : /Templates/Slack/

If you plan to use a template that is not specific should be stored under generic.

File: asabiris/orchestration/sendslack.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mithunbharadwaj) [reply asabiris/orchestration/sendslack.py]: @eliska-n it is /Templates now.

[the remark this answers is not in the document]
Comment (mithunbharadwaj) [reply asabiris/orchestration/sendslack.py]: @eliska-n New response 

`{
    "result": "ERROR",
    "message": "Templates must start with /Templates",
    "uuid": "63ec3c74-f5cf-4a9d-b8d1-fcb28a2853f0"
}`

File: asabiris/orchestration/render.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (mithunbharadwaj) [reply asabiris/orchestration/render.py]: @eliska-n New response when template won't start with /Template and it is an absolute path 

`{ "result": "ERROR", "message": "Templates must start with /Templates", "uuid": "63ec3c74-f5cf-4a9d-b8d1-fcb28a2853f0" }`

File: asabiris/orchestration/sendmail.py (hunk 10 above)
Comment (mithunbharadwaj) [reply asabiris/orchestration/sendmail.py]: @eliska-n is the order of picking up the subject "request > template > config?

File: asabiris/orchestration/sendmail.py [hunk 11]
```diff
@@ -119,6 +125,9 @@ async def render(self, template, params):
 
 		jinja_output will be used for extracting subject.
 		"""
+		if template.startswith("/"):
```
Comment (eliska-n) [inline asabiris/orchestration/sendmail.py:128]: My testing library does not contain "/Templates" directory at all and the e-mail still proceeds.

```python
if template.startswith("/"):
    assert template.startswith("/Templates"), "Your template must be stored in /Templates directory"
else:
    template = "/Templates/" + template
```

@ateska This is what we agreed on during review. However, as I saw during demo, further breakdown of the Templates folder (Mails, Slack etc...) is being prepared together with asab-library, right?
It only makes my feeling about the necessity of using absolute paths stronger.
Just now, I cannot really decide, what the else branch should look like.

Comment (mithunbharadwaj) [reply asabiris/orchestration/sendmail.py:128]: @eliska-n Yes , because iris accepts relative paths.

**Pull request merged**